### DISCUSSION TITLE
Is there a way to restrict `FieldPath` to only allow ones pointing to a certain type?

### DISCUSSION THREAD

**Discussion by Svish** [react-hook-form/react-hook-form#8550] [Q&A]
I have a `InputCheckboxGroup` component, which I use like this:

```ts
<InputCheckboxGroup
  control={control}
  name="favoriteFruites"
  options={['Apple', 'Banana', 'Pear', 'Pineapple']}
/>
```

The implementation is below, it works, and the `name` gets restricted by what the passed `control` is typed to allow. So in this case, there will be a type error if `favoriteFruites` is not a valid path in the form.

But, is there a way to _also_ limit this `name` to only allow paths pointing to for example a `string[]` value? My `InputCheckboxGroup` component expects the value to be an array of strings, but the given `name` could point at anything. So when I try to call `setValue` in the component (see comment in code below), then I get this error:

> Argument of type 'never[]' is not assignable to parameter of type 'UnpackNestedValue<PathValue<TFieldValues, TFieldName>>'.

Is there a way I to type this component, and the `name` prop in particular, so both the `value` from `useController` and the `value` parameter of `setValue` from `useFormContext` knows they are an array of `string[]`?

```ts
interface Props<
  TFieldValues extends FieldValues = FieldValues,
  TFieldName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
> {
  name: TFieldName; // This successfully typechecks the name, but doesn't restrict which type it points to
  control: Control<TFieldValues>;
  options: readonly string[];
}

export default function InputCheckboxGroup<
  TFieldValues extends FieldValues = FieldValues,
  TFieldName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
>({ name, control, options }: Props<TFieldValues, TFieldName>): ReactElement {
  const {
    field: { value, onChange, onBlur },
  } = useController<TFieldValues>({ name, control });
  const [checked, setChecked] = useState<string[]>(value ?? []);

  const { setValue } = useFormContext<TFieldValues>();

  useEffect(() => {
    if (!Array.isArray(value)) {
      setValue(name, []); // <-- fails, because [] is not a valid value according to 
    }
  }, [name, value, setValue]);

  return (
    <>
      {options.map((value) => (
        <input
          type="checkbox"
          key={value}
          value={value}
          checked={checked.includes(value)}
          onChange={(e) => {
            const set = new Set(checked);
            if (e.target.checked) {
              set.add(e.target.value);
            } else {
              set.delete(e.target.value);
            }
            onChange([...set]);
            setChecked([...set]);
          }}
          onBlur={onBlur}
        />
      ))}
    </>
  );
}
```

**Comment by danieledelgiudice** [comment]
Have you managed to find a solution? I am struggling with the same issue.

**Reply by Svish** [reply]
Nope.

**Reply by danieledelgiudice** [reply]
It appears a PR has been recently accepted with a new type for this use case: https://github.com/react-hook-form/react-hook-form/pull/8837.
Looking forward the release 🤞

**Reply by Svish** [reply]
Oooooh, that looks very interesting, thanks for letting me know. Will definitely keep a look out for that in new releases!
